src/main/java/com/microsoft/azure/relay/AsyncSemaphore.java

```diff
@@ -4,17 +4,20 @@
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionException;
 import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Supplier;
 
 public class AsyncSemaphore {
 	private final Object thisLock = new Object();
+	private final ScheduledExecutorService executor;
 	private final int limit;
 	private InputQueue<Boolean> waiterQueue;
 	private int permits;
 	
-	AsyncSemaphore(int permits) {
+	AsyncSemaphore(int permits, ScheduledExecutorService executor) {
```

Comment:
>executor [](start = 54, length = 8)

It's better to put this in the .ctor because when the other calls were providing their own they might have actually used one from a previous call. Better to be consistent instead of misleading.